### ISSUE TITLE
Rewrite our github history to remove commit related to the original rust repository

### ISSUE THREAD

**Issue by celinval** [model-checking/kani#991]
Our project started as a fork of the rust compiler which we kept up-to-date with the rust repository. We no longer require any modification to the compiler code and we use their libraries as is. However, our commit history still includes all previous commit to the rust repository that precedes our architecture modification.

We should consider pruning our repository to only include commits that are relevant to the project. The advantage would be to provide a more meaningful history, which is easier to navigate and that is also faster to clone.

**We should first assess if the cost is worthwhile.** Users that have already cloned their repository will have to recreate their repository in order to track the new history. Any PR or link to old commit will be invalidated.

- [ ] Create a fork of our project.
- [ ] Copy a few issues and PRs (open and closed) to the forked project.
- [ ] Prune the forked project history.
- [ ] Assess the impact.

If we decide to move on:
- [ ] Create a "snapshot" of our project with the current history and ALL issues and PRs.
- [ ] Prune the original repository

To do the pruning itself, it looks like `git` [no longer recommends](https://git-scm.com/docs/git-filter-branch#_warning) using `git filter-branch`. One possibility is to use the tool linked in the git doc: [git-filter-repo](https://github.com/newren/git-filter-repo).

To copy issues and PRs, we might need to use the GitHub REST APIs: https://docs.github.com/en/rest/reference

**Comment by tedinski** [comment]
I don't think we'd worry about copying issues. This repo should continue to be the one we work out of, and we should retain the old commit history in a forked repo.

**Comment by celinval** [comment]
Yeah, I agree that copying the issues is not a must have. It might be nice to do as part of "backing up" the project.

**Comment by tedinski** [comment]
I got nerd-sniped on this a little.

Plan, instead of `filter-repo`:

0. Push current history to another repo
1. `git rebase -Xours origin/upstream-rustc` This *crudely* (`-Xours` on any merge conflict) rewrites history so all our commits are on top of the last time we merged with rustc. This "breaks" our history, probably, a little. `-Xours` is going to just slam-resolve things in a way that means old commits likely won't build anymore. But many of them didn't already because of our old rebasing strategy.
2. Truncate our git history to this first commit, thus pruning all of the rustc history.
3. force push this as `main`

Things will probably be a bit messed up as a result, but I don't think hugely. I can try to create a test repo to preview this, soon.

I couldn't find an approach to doing this sort of thing with filter-repo.

before:

```
635M	.git/objects
```

after rewrite and gcing everything:

```
302M    .git/objects
```

I think `objects` is the one to compare, and it's pretty big still, but that's because our initial commit is still the whole rustc compiler source. So we're losing the history (big), but not all the contents of those files (still big).

That's troubling, we're not getting as much savings as we'd like.

## Saving PRs

after we rewrite, we should be able to save PRs with `rebase --onto`

Basically, read `git rebase --onto new-parent first-commit`

So if I have a PR with two commits (old_main <- A <- B) I:

```
# get new commits without changing local branches
git fetch origin
# work on my PR
git switch my-pr
# find first commit id of my pr, where it was based off main
git log
# rebase
git rebase --onto origin/main that-commit-id
```

There might be a way to simplify this, but that's my current understanding.

**Comment by tedinski** [comment]
```
git filter-repo --invert-paths --path src/test/ui
du -h -d1 .git
27M     .git/objects
```

That seems entirely too good lol.

Well, with just one edit, maybe this IS worthwhile!

**Comment by tedinski** [comment]
Proposed rewrite script:

```bash
#!/bin/bash

set -eu

# repo where kani is checked out (mine is still called rmc)
ORIGINAL=rmc
NEW=$(mktemp -d rewritten-history-XXXXXXXX)

echo "Working: $NEW"

validate_repo() {
    diff -ru -x .git $ORIGINAL $1 
}

## 1. Clone and double check we're good to go

git clone $ORIGINAL $NEW/first
(cd $NEW/first && git submodule update --init)

if ! validate_repo $NEW/first; then
    echo "Cloned repo is not identical."
    echo "Uncommitted changes? Or ignored files? Try:"
    echo "  git clean -xffd"
    echo "  git submodule foreach --recursive git clean -xffd"
    exit 1
fi

## 2. Purge the 'rustc' history

# Let's try git replacing/grafting all merge commits to edit out the rustc parent from history.

# This is a list of all our merges with rustc, editing out the parent commits from rustc (by preserving only our commits)
cd $NEW/first
git replace --graft f66fad280647260478e6541b2a834effa7d97a9d 8f3d0911697
git replace --graft d7bda4ff90b3ac8298d0e1722275c19bf0b55716 f7c13ea47a8
git replace --graft 414dca40e6be76ce0e323ee5a66db756bee58b63 2fb5bd16d91
git replace --graft 5a27a70a7a924d8cd63cfd0a6c49d7cd9163e760 625990c8a5c
git replace --graft 7d0d4cfc6da58b0f09575e93085a771bf69c0774 95356a0e9b4
git replace --graft d6e8fd7e7da5d941294f1aeee4bd2e06b66633ae 9a8f75bf2c1
git replace --graft 2141a76f98dbe3e339c6a1e184ac778965f2da54 52835ca0d1d
git replace --graft 91c9c070350e61d1546bacd2868d428e6f5bdfb5 5d23cf86665
git replace --graft 92e82d7182dc88071af94a6d0e41ce635e906f6a 805299f84a4
git replace --graft 1081e525bbabde7b4ed5e1f3ea0e8f73d16115e9 a728d8d41a1
git replace --graft 4001d5100d560becf6e6f9084a6783824f6e5fee 7635cfebfa5
git replace --graft 6f6d3602194fe5a54e21482ceaa95292bf90d123 066d22ac385
git replace --graft a11473fd14fc40964cf1f915522db5cd256a9403 72d7cd41243
git replace --graft a9d1eb2dcbe4820a95e45b95673b66463fd69d81 77f3cd015d6
git replace --graft 0764e2a8bf4f443a112305995159c1c7b5fba6cd dff61386585
git replace --graft 38d6e6327361efa68f6a5b6d952f33bc3c5513dc d49f4c9925e
git replace --graft 7d8413fbb2c3906aeb30cba04cf77396f1593fb6 898906c420f

# purge the parent of the "initial rmc commit" (so: no parent of this commit)
git replace --graft c9479ff70251d3148eb0213b6144f201595ac3fa
cd -

## 3. History apocalypse

# 'git replace' is kind of a temporary thing. It doesn't actually rewrite history, it just pretends.
# So this action does two things:
# 1. Remove all those UI tests, which are the majority of the rustc project size and we never cared about.
# 2. Make the "let's pretend" real.

# oh, but quickly remove extra branches besides main, filter-repo will pay attention to branches from fetched remotes apparently
(cd $NEW/first && git remote remove origin)

(cd $NEW/first && git filter-repo --force --invert-paths --path src/test/ui)

validate_repo $NEW/first

echo "new size:"
du -h -d0 $NEW/first/.git/objects
# Looks to be 30 MB, down from 660 MB

## 4. Rewrite the commit message on the original commit

# there are simpler interactive ways of doing this (git rebase -i --root)
# but this was the way I found to do it in a script / from the command line only

INITIAL_COMMIT=$(cd $NEW/first && git rev-list --max-parents=0 HEAD)

NEW_MESSAGE=$(cat <<END
Truncated history.

Kani (originally 'RMC') used to be a fork of Rust.
This later changed, and this history truncation is the result of filtering out
Rust's history and some of its files from the Kani repo, to reduce repo size.
Some more information about this truncation can be found here:
https://github.com/model-checking/kani/issues/991

RMC originally pursued a rebase strategy for syncing with rustc and then
switched to merging, before finally using rustc as a library.
When this git history was truncated, this commit had rustc parent:
https://github.com/rust-lang/rust/tree/f8810ee171d1549cfb4082d4e1289704aad722e9
This commit reflects the last time the rebase strategy was applied.
END
)

cd $NEW/first
git checkout --orphan new_history $INITIAL_COMMIT
git commit -m "$NEW_MESSAGE"
git rebase --onto new_history $INITIAL_COMMIT main
cd -


## 5. double check

validate_repo $NEW/first

echo
echo "Done rewriting history. End of my script; success."
echo $NEW
echo
```

The result:

https://github.com/tedinski/rmc/commits/new_history

**Comment by celinval** [comment]
Nice! I would suggest a few tweaks.

1. Start from the original repository: `git@github.com:model-checking/kani.git`.
2. Add `--committer-date-is-author-date` to the rebase so the author dates are used.
3. Use Kani instead of RMC in the second paragraph.

**Comment by tedinski** [comment]
2 & 3 sound good.

What do you mean by 1?

**Comment by celinval** [comment]
This line:
```
ORIGINAL=rmc
NEW=$(mktemp -d rewritten-history-XXXXXXXX)
```

I would replace by something like:
```
ORIGINAL=$(mktemp -d original-history-XXXXX)
git clone git@github.com:model-checking/kani.git $ORIGINAL
(cd $ORIGINAL && git submodule update --init)

```

**Comment by tedinski** [comment]
History rewritten. There are a few remaining followups to note:

* [ ] Push old history to an archive repository somewhere
* [ ] Delete old branches
* [ ] Rebase the closed pull requests (1 done, 4 to go)
